pkg/cloud/openstack/openstack.go

```diff
@@ -80,3 +84,73 @@ func NewProviderAssets(config config.OperatorConfig) (common.CloudProviderAssets
 
 	return assets, nil
 }
+
+// CloudConfigTransformer implements the cloudConfigTransformer. It uses the source ConfigMap and
+// to create a new config that adheres to the configuration for CCM
+// It returns an error if the platform is not OpenStackPlatformType.
+func CloudConfigTransformer(source *corev1.ConfigMap, infra *configv1.Infrastructure) (*corev1.ConfigMap, error) {
+	if !(infra.Status.PlatformStatus != nil &&
+		infra.Status.PlatformStatus.Type == configv1.OpenStackPlatformType) {
+		return nil, fmt.Errorf("invalid platform, expected to be OpenStack")
+	}
+
+	output := source.DeepCopy()
+	infraConfigKey := infra.Spec.CloudConfig.Key
+	if val, ok := output.Data[infraConfigKey]; ok {
+		output.Data[common.DefaultConfigKey] = val
+		delete(output.Data, infraConfigKey)
+	} else {
+		return nil, fmt.Errorf(
+			"key %s specified in infra resource was not found in source configmap %s",
+			infraConfigKey, client.ObjectKeyFromObject(source),
+		)
+	}
```

Comment:
Every transformer is currently doing this. The target CM and key are common in all cases. The CloudConfigTransformer should take a *string* as input containing the input config from the relevant configmap[key], and write a transformed *string*. The parent code should then write that to the target configmap[DefaultConfigKey], which will be the same in all cases(?).

For input we should have 2 cases:
* We're using CCO
* We're doing the transformation in CCCMO

If we're using CCO we should never fall back to the user-managed config. If CCO hasn't done its thing yet we wait until it has. For the first cut we should leave AWS and Azure configs in CCO, but it's good to play with them at the WIP stage so we can increase our confidence that we *can* implement them here.

If we're doing the transformation in CCCMO the source should always be the configmap[key] referenced in infra.

I think this would make the return from GetCloudConfigTransformer something like:

```golang
type CloudConfigTransformer func(config string, infra *configv1.Infrastructure) (string, error)

func GetCloudConfigTransformer(platformStatus *configv1.PlatformStatus) (CloudConfigTransformer, error)
```

If GetCloudConfigTransformer returns nil for the transformer we use CCO as the source CM and the identity transformer. If it returns an actual transformer we use the user config as the source. In both cases the same code writes the returned value to the same destination.

Reply:
I've done just this, I think. Let me know if I missed anything.